### ISSUE TITLE
Quest helper like behavior in bank

### ISSUE THREAD

**Issue by DapperMickie** [geel9/runelite-hunter-rumours#70]
Make a bank tab/quest helper like icon to help you get all the items you need for hunter rumours. With a toggle to show all/show for your current rumour.

**Comment by DapperMickie** [comment]
# Issue 70: Hunter Rumours bank helper

## What the issue is asking for

[Issue #70](https://github.com/geel9/runelite-hunter-rumours/issues/70) asks for a Quest Helper-style button in the bank. Opening it should show the gear needed for either the current rumour or every rumour.

The important word is "Quest Helper-style." A normal bank search can only filter items that already exist in the bank. Quest Helper also draws missing items, quantities, section headings, and status marks. Matching that behavior requires a synthetic bank layout, not just a search string.

## Recommended item catalog

The catalog should distinguish required items from useful items. Otherwise the all-rumours view turns into an unhelpful shopping list of every conceivable teleport and optimisation.

The OSRS Wiki's current [Hunters' Rumours strategy guide](https://oldschool.runescape.wiki/w/Hunters%27_Rumours/Strategies) provides a good baseline for both groups. Its method table agrees with the game's general [Hunter documentation](https://oldschool.runescape.wiki/w/Hunter): pitfall traps need a knife, logs, and a teasing stick or hunter's spear; net traps need ropes and small fishing nets; deadfalls need a knife and logs; kebbit tracking needs a noose wand; falconry needs an access payment rather than a bankable glove requirement.

### Required by method

| Method | Rumours in this repository | Items to show | Quantity and alternatives |
| --- | --- | --- | --- |
| Bird snare | Tropical wagtail | Bird snare | Dynamic trap count, capped at 5 |
| Box trap | Embertailed jerboa, grey chinchompa, red chinchompa | Box trap | Dynamic trap count, capped at 5 |
| Net trap | Swamp lizard, orange salamander, red salamander, tecu salamander | Small fishing net and rope | One matched pair per available trap |
| Deadfall | Wild kebbit, barb-tailed kebbit, prickly kebbit, sabre-toothed kebbit, pyre fox | Knife or fletching knife, plus any logs | One knife alternative and at least one log. Showing several logs is useful but should remain a recommendation |
| Spiked pit | Spined larupia, horned graahk, sabre-toothed kyatt, sunlight antelope, moonlight antelope | Knife or fletching knife, any logs, teasing stick or hunter's spear | One of each alternative group |
| Butterfly or moth | Sapphire glacialis, snowy knight, black warlock, sunlight moth, moonlight moth | Butterfly net or magic butterfly net when needed | Moths are unlocked at a level where they can be caught bare-handed. A net is only strictly required for the lower-level butterfly rumours. Butterfly jars are optional |
| Kebbit tracking | Razor-backed kebbit | Noose wand | One |
| Herbiboar | Herbiboar | Nothing | No required bank item |
| Falconry | Spotted kebbit, dark kebbit, dashing kebbit | Coins for access | Show 500 coins unless the permanent 500,000-coin access unlock can be detected. Do not show a falconer's glove as a bank withdrawal requirement |
| Goat pit | Wyrmscraig goat | Cattleprod | One, based on this repository's current `GOAT_PIT` mapping. This new method is not yet present in the cited strategy table and deserves an in-game verification pass before release |

The dynamic trap count follows the game's trap limits: 1 at level 1, then one additional trap at levels 20, 40, 60, and 80, for a normal maximum of 5. The [Hunter page](https://oldschool.runescape.wiki/w/Hunter#Multiple_traps) documents those thresholds. Use the player's real Hunter level and cap the result to the number relevant to the selected method.

The existing model cannot represent this table. [`Trap.java`](https://github.com/geel9/runelite-hunter-rumours/blob/a08cfd2b057fdff54f8295add91d06d560c66a1f/src/main/java/com/geel/hunterrumours/enums/Trap.java) stores one item ID per method. That already omits rope for net traps, knife for deadfalls and pits, logs for pits, item alternatives, quantities, and the fact that Herbiboar needs no tool. [`Creature.java`](https://github.com/geel9/runelite-hunter-rumours/blob/a08cfd2b057fdff54f8295add91d06d560c66a1f/src/main/java/com/geel/hunterrumours/enums/Creature.java) is still useful as the creature-to-method mapping.

### Useful for most or all rumours

Show these in a separate "Recommended" section so users can tell them apart from blockers:

- All four guild hunter outfit pieces: headwear, top, legs, and boots. The full set raises catch rate by 2.5% and rare-part chance by 5%, according to the [strategy guide](https://oldschool.runescape.wiki/w/Hunters%27_Rumours/Strategies#Equipment). The repository already uses these four item IDs when calculating worn outfit pieces in [`HunterRumoursPlugin.java`](https://github.com/geel9/runelite-hunter-rumours/blob/a08cfd2b057fdff54f8295add91d06d560c66a1f/src/main/java/com/geel/hunterrumours/HunterRumoursPlugin.java).
- Quetzal whistle, hunter cape, or max cape as return-to-guild alternatives. Prefer whichever owned variant the resolver finds.
- Huntsman's kit.
- Meat pouch and fur pouch variants.
- Bonecrusher or bonecrusher necklace.
- Weight-reducing clothing or ring of endurance.

Useful method-specific items belong in the same optional section: butterfly jars; an axe and log basket for deadfalls and pits; magic secateurs and herb sack variants for Herbiboar; ring of pursuit for kebbit tracking; and a chisel for antelope antlers or kebbit spikes. These all come from the strategy guide's method table.

I would keep arbitrary travel teleports out of the first release. The repository already owns location and Shortest Path behavior, while travel inventories vary by preferred location, spellbook, construction unlocks, and account type. A later version can derive travel suggestions from `RumourLocation` preferences rather than hard-coding another parallel location table.

## Data model to add

Add a small requirement model independent of `Trap`:

```text
HunterGearRequirement
  name
  item alternatives
  quantity rule
  required or recommended
  applies to methods and/or individual creatures
  optional ownership condition
```

`HunterGearCatalog` should return resolved sections for a `Rumour` and player state. The resolver should:

1. Add global recommended gear.
2. Add method and creature requirements.
3. Calculate trap quantities from the real Hunter level.
4. Choose an owned ID among alternatives for display, falling back to a canonical ID.
5. Merge duplicate requirements in the all-rumours view. Alternative groups must stay groups, not become separate mandatory rows.
6. Check bank, inventory, and equipped containers when deciding whether a requirement is satisfied.

For `CURRENT`, resolve `plugin.getCurrentRumour()`. If it is `Rumour.NONE`, keep the tab button visible but show an explanatory empty state. For `ALL`, resolve `Rumour.allValues()`, group by trapping method, then add the common recommended section once. "All" should mean all supported rumours, not only the rumours currently stored against enabled guild hunters.

## Bank interface architecture

Quest Helper is the closest proven implementation. Its [`QuestBankTabInterface`](https://github.com/Zoinkwiz/quest-helper/blob/633ab56e2eb3eb363f21da3fd75f6f2bc0fa073a/src/main/java/com/questhelper/bank/banktab/QuestBankTabInterface.java) creates a child graphic on the bank's root widget, toggles a selected sprite, resets `BankSearch` to trigger layout, and closes itself when the player selects another bank tab, a tag tab, potion storage, or search. Its [`QuestBankTab`](https://github.com/Zoinkwiz/quest-helper/blob/633ab56e2eb3eb363f21da3fd75f6f2bc0fa073a/src/main/java/com/questhelper/bank/banktab/QuestBankTab.java) keeps the bank in a searching state, hides the normal item widgets, lays out real and synthetic items, and rewrites menu indexes so withdrawals still target the real bank container. Its [`QuestHelperBankTagService`](https://github.com/Zoinkwiz/quest-helper/blob/633ab56e2eb3eb363f21da3fd75f6f2bc0fa073a/src/main/java/com/questhelper/bank/banktab/QuestHelperBankTagService.java) resolves alternatives and selects an owned item ID for display.

The Hunter Rumours implementation can be much smaller because it does not need quest stages, Grand Exchange search replacement, keyrings, or potion storage in the first release. It still needs these pieces:

- `HunterBankTabInterface`: owns the bank button, selected state, `CURRENT` or `ALL` mode, and clean activation/deactivation.
- `HunterBankLayout`: turns resolved requirements into headings, real bank item widgets, missing-item widgets, quantities, and optional check marks.
- `HunterGearCatalog`: pure requirement data and resolution logic.
- A compact `BankTabSection` and `BankTabItem` view model.

Use a config enum such as `BankHelperMode.CURRENT` and `ALL`, plus an `enableBankHelper` setting. The button's context menu should also expose "Show current rumour" and "Show all rumours" so the user can switch without closing the bank. Persist the last choice in config.

## RuneLite event and script hooks

The minimum event set, based on Quest Helper and RuneLite's own bank implementation, is:

- `WidgetLoaded` for `InterfaceID.BANKMAIN`: create the button and restore its active state.
- `WidgetClosed` for `InterfaceID.BANKMAIN`: clear widget references and transient layout state.
- `MenuOptionClicked`: deactivate on another native tab, tag tab, search, or potion store selection; remap withdrawal actions from displayed real items to their actual bank indexes.
- `ScriptPreFired` for `ScriptID.BANKMAIN_FINISHBUILDING`: restore any widget sizes changed by the synthetic layout and set the custom title while active.
- `ScriptPostFired` for `ScriptID.BANKMAIN_FINISHBUILDING`: remove previously added children and lay out the current resolved sections after the client has rebuilt the bank.
- `ScriptPostFired` for `ScriptID.BANKMAIN_SEARCHING`: force the return value to true while the custom tab is active.
- `ScriptCallbackEvent` for `getSearchingTagTab`: report the custom tab as a tag-like search layout.
- `ScriptPreFired` for `ScriptID.BANKMAIN_SEARCH_TOGGLE`: close the custom tab before normal search opens.
- `ItemContainerChanged` for bank, inventory, and equipment: refresh status and quantities when ownership changes.
- Existing rumour/config changes: refresh the active layout when the current rumour or mode changes. `refreshAllDisplays()` is the natural integration point, but bank refresh should live behind a dedicated method so unrelated display refreshes do not rebuild a closed bank.

RuneLite's [`ScriptID`](https://github.com/runelite/runelite/blob/master/runelite-api/src/main/java/net/runelite/api/ScriptID.java) documents the bank build, finish-building, searching, and search-toggle scripts. RuneLite's [`BankPlugin`](https://github.com/runelite/runelite/blob/master/runelite-client/src/main/java/net/runelite/client/plugins/bank/BankPlugin.java) also defers `BankSearch.layoutBank()` on search refresh because the vanilla client otherwise relays out periodically rather than on every input change.

All widget mutation and `BankSearch` calls must run on the client thread. Startup should register or initialise the bank helper. Shutdown must close the synthetic tab, reset search, restore original children and dimensions, and remove event registrations. An incomplete cleanup path can leave the bank visually broken until it is reopened.

## Rendering behavior

For each item row:

- Reuse the real bank item widget when the item exists in the bank. This preserves ordinary withdraw actions.
- Draw a desaturated synthetic item when it is absent from the bank. Give it a non-withdraw action such as "Details" and a tooltip explaining the requirement.
- Show the required quantity and a check or cross based on the combined bank, inventory, and equipment count.
- Resolve noted IDs and placeholders to their canonical item ID before comparing ownership.
- Label sections. In current mode, "Required" and "Recommended" are enough. In all mode, use "Common", then trapping-method sections.

The first release should treat potion storage as out of scope. If stamina potions or other stored potions are later added to recommended gear, Quest Helper's potion-store adapter shows why this is its own feature: potion storage is a separate widget/container and withdrawal target, not part of the normal bank item array.

## Risks and decisions

1. **Bank scripts are brittle.** The implementation depends on client scripts and widget geometry. Pin tests to the current RuneLite API and manually test fixed, resizable, tag tabs, placeholders, rearrange and swap modes, search, potion storage, and bank reopening.
2. **Plugin interaction.** Bank Tags, Bank Tag Layouts, Inventory Setups, Wasted Bank Space, and Quest Helper can all alter the same item container. The custom tab must only rewrite widgets while active and restore them before each vanilla rebuild.
3. **Withdraw correctness.** Duplicate items and synthetic widgets make displayed indexes differ from `InventoryID.BANK` indexes. Remapping by item ID is mandatory. A missing row must never send a withdrawal action.
4. **Item alternatives.** Knife versus fletching knife, butterfly net variants, whistle variants, pouches, capes, and hunter's spear versus teasing stick must count as OR requirements. Flattening these creates false missing-item warnings.
5. **Equipped outfit.** A player can be wearing all four pieces while viewing the bank. The layout should mark those requirements satisfied rather than claim they are missing.
6. **Falconry state.** The glove is supplied by Matthias and the actual requirement is the fee or permanent unlock. If the unlock cannot be read from a stable varbit, show the 500-coin fee with wording that it may already be unlocked.
7. **Wyrmscraig goat data.** The repository maps Goat Pit to a cattleprod, but the public strategy source used here has not caught up. Confirm the required item and whether other supplies are involved in-game before locking the catalog.
8. **Future rumours.** Add a coverage test so every non-`NONE` rumour resolves to a known method rule, including a deliberate no-item rule for Herbiboar. This makes new enum entries fail loudly instead of silently disappearing from the bank helper.

## Implementation phases

### Phase 1: catalog and resolver

Add the requirement types, complete mappings, current/all mode, dynamic quantities, ownership checks, and unit tests. No widgets yet. This is the safest place to settle what "all items" means.

### Phase 2: read-only bank tab

Add the bank icon, mode switch, headings, real item display, and synthetic missing rows. Missing rows get "Details" only. Refresh on rumour and container changes. This delivers the issue's core value.

### Phase 3: withdrawal-safe polish

Remap real item menu indexes, handle placeholders and noted IDs, add quantity/status marks, and harden cleanup across other tabs, search, potion storage, logout, plugin disable, and bank reopen.

### Phase 4: optional improvements

Add Grand Exchange search for missing tradeable items, travel-item suggestions derived from preferred locations, potion storage integration, and user controls for recommended categories. These should not block issue #70.

## Test plan

Unit tests should cover catalog completeness, every method mapping, alternative-item satisfaction, dynamic trap quantities at threshold levels, deduplication in all mode, `Rumour.NONE`, equipped outfit detection, and current-to-all mode switching.

Manual client tests should cover opening and closing the custom tab, changing rumours while it is open, withdrawing a real item, clicking a missing item, empty and placeholder-only banks, duplicate IDs, native tabs, tag tabs, search, potion storage, fixed and resizable layouts, rearrange and swap modes, logout, plugin disable, and coexistence with Quest Helper and Bank Tag Layouts.

## Bottom line

The clean implementation is a dedicated requirement catalog plus a small Quest Helper-derived synthetic bank layout. Trying to stretch the existing single `Trap.ItemId` field or using a normal bank search will miss multi-item methods, alternatives, quantities, equipped gear, and absent items. Phases 1 and 2 are the meaningful MVP. Phase 3 is what makes it safe enough to ship broadly.
